### ISSUE TITLE
Is there support for Hot Module Replacement?

### ISSUE THREAD

**Issue by jared-dykstra** [Microsoft/redux-dynamic-modules#53]
Two questions:
1.  Is `replaceReducer()` supported by the store, and is there a way to substitute a saga?
1.  Is there any guidance (via example code or docs) to show how to correctly setup for HMR in reducer and/or saga code?

**Comment by abettadapur** [comment]
We don't have any HMR support right now, but it is something that we could invest in.
For your other questions, where does the module system fall short in accomplishing these goals?

**Comment by jared-dykstra** [comment]
> where does the module system fall short

I don't think it does.  I opened this PR to ensure that I'm using it correctly and making the most of it--I wasn't sure.  I figured, if I had missed something, if nothing else I could PR against the documentation.

So far, I'm quite impressed with the library and using it in a CRA v2 project.

**Comment by jared-dykstra** [comment]
For example, I was expecting to have to setup Redux Devtools as an extra step, but was surprised to discover that it "just worked"

**Comment by abettadapur** [comment]
With regards to `replaceReducer`, we do not expose that function. Instead, you can use modules to add/remove reducers. 

Are you asking for a function like `replaceModule`, which could replace a module at runtime? We dont have that today, but it is something we could potentially add to support hot reloading scenarios

**Comment by jared-dykstra** [comment]
Here's a concise description of the problem I'm seeing.  I suspect this is user-error (an error in my code)   I currently have this code in place for HMR, and my app is using react-router:

```js
renderRootNode(() => <App {...props} />)

if (process.env.NODE_ENV !== 'production') {
  if (module.hot) {
    module.hot.accept('./App', () => {
      // eslint-disable-next-line no-console
      console.warn('[HMR] - Updating App')
      // eslint-disable-next-line global-require
      const NextApp = require('./App').default
      renderRootNode(() => <NextApp {...props} />)
    })
  }
}
```

When I make a change, HMR kicks in.  But then I notice two dispatched redux actions:

1. `@@router/LOCATION_CHANGE` - I don't expect this to fire, but it does.  The location is the same as it was before, and the state diff seems inconsequential.  I think I can ignore this event.
1. `@@Internal/ModuleManager/ModuleRemoved` - Wipes out all state corresponding to this module
1. `@@Internal/ModuleManager/ModuleAdded` - Sets up the same module again from its initial state

Of course, this only affects development, but I'd love to avoid losing the module's state every time a development change is made.  If some changes need to be made via a PR, I'm happy to look into it, but am not sure if there's a way.

So, `<App />` uses react-router, which eventually renders `<Page />` which looks like this:

Page.js:
```js
const PageDynamic = React.lazy(async () =>
  import(/* webpackChunkName: 'PageDynamic' */ './PageDynamic')
)

<Suspense fallback={...}>
  <PageDynamic />
</Suspense>

```

PageDynamic.js:
```js 
import { DynamicModuleLoader } from 'redux-dynamic-modules'

import { module } from 'reduxStore/myModule'
import RealPage from './RealPage'

const PageDynamic = props => (
  <DynamicModuleLoader modules={[module()]}>
    <RealPage {...props} />
  </DynamicModuleLoader>
)
```

Lastly, `<RealPage />` is a regular react component, rendering the state of `reduxStore/myModule`

**Comment by winstonjz** [comment]
HMR would be killer for this project. 👍

**Comment by jared-dykstra** [comment]
The way I see it, there are a few things to do:
1. Each module needs to have it's own `module.hot.accept()` added to it, listening for changes in it's own dependencies.  Traditional use of `module.hot.accept('./something')` relies on a full chain of dependencies from `./something` -- But that would create a dependency chain and defeat the purpose of this redux-dynamic-modules
1. A process like: `saveState()`, `createStore(prevState)`, `restartSagas()`   ... Or, use of redux's `replaceReducers()` to avoid creating a new store.   I'm not sure which option is best.  Terminating and restarting sagas could lead to some indeterminism, but should be ok?  Ideally, this process would apply to each module individually to have a minimal effect.

To be clear: I think this library is fantastic.  I'm totally willing to help out with a PR, but would love to know what approach you would suggest.

Can you offer any advice?

**Comment by jkomyno** [comment]
Hi @abettadapur and @jared-dykstra, any update on this yet?

**Comment by navneet-g** [comment]
@jared-dykstra I can take a look, where do you see code for option 1 or 2 being called? I am not familiar with how runtime detects that code is being hot reloaded, please educate me a bit more.

**Comment by navneet-g** [comment]
@jared-dykstra I can take a look, where do you see code for option 1 or 2 being called? I am not familiar with how runtime detects that code is being hot reloaded, please educate me a bit more.

**Comment by navneet-g** [comment]
@jared-dykstra ... so I did some reading, Option 2 above seems better to me, for the reasons you mentioned and it is very explicit. Ideally the ModuleManager should handle replacing reducer and extensions/middlewares/sagas. We need to find a way so ModuleManager can be notified that files belonging to a rdm-module is being reloaded. Any suggestions?

**module** : module in terms of a javascript module/file.
**rdm-module**: A module as defined by redux-dynamic-modules.

**Comment by navneet-g** [comment]
Ok ... further reading, this is what we can do.
1) When a module detects it is hot reloaded, it dispatches a global event.
2) The ModuleManager listens to such events, gets the moduleid and module contents from the event and replaces the module while retaining the state and not dispatching initial/final actions

**Comment by shinvey** [comment]
If you use [react hot loader](https://github.com/gaearon/react-hot-loader) like me, there is a workaround that could help you.
```js
function yourComponent() {
  /**
   * There are two options important to help react hot loader work properly in common cases
   * `module.hot.active` can make sure DynamicModuleLoader get updated rdm-module not old one
   * `strictMode` to ensures that DynamicModuleLoader remove rdm-module before add rdm-module
   */
  return module.hot && module.hot.active && <DynamicModuleLoader strictMode={true} modules={[getHackerNewsModule()]}>
    <ConnectedHackerNews />
  </DynamicModuleLoader>
}
```
Um... I'd like to say it really was my tough time to debug code between react-hot-loader and DynamicModuleLoader that I need to figure out how they work together.
